### PR TITLE
Add AFL in zcutil (with all-in-one script)

### PR DESCRIPTION
**Pull request by defuse**
Supersedes #4156 and #4167.

Fuzzing targets and input sets are defined by the contents of directories in `./src/fuzzing/`. Inside the directory, there's a `fuzz.cpp` and `fuzz.h` with a `main()` function that will replace `zcashd`'s actual `main()` as well as an `input` subdirectory containing the inputs, one per file. To just run a fuzzer, you can, for example...

```
make clean # if you've previously build zcashd without AFL instrumentation
./zcutil/afl/afl-getbuildrun.sh DecodeHexTx
```

Alternatively you can...

```
./zcutil/afl/afl-get.sh /tmp/afl   # (or wherever you want to build AFL)
./zcutil/afl/afl-build.sh /tmp/afl DecodeHexTx -j$(nproc)
./zcutil/afl/afl-run.sh /tmp/afl DecodeHexTx
```

Run `make clean` whenever you switch between a normal build and an AFL-instrumented build.

### CODE REVIEW

**Comment by defuse** [comment]
@zebambam thanks for the suggestions, I updated the PR.

**Comment by zebambam** [comment]
Legend!

**Comment by zebambam** [comment]
@homu try

**Comment by zebambam** [comment]
@zkbot try

**Comment by zebambam** [comment]
Small point because this is a plumbing PR, but the one that I tested isn't finding new paths..

`                       american fuzzy lop 2.52b (zcashd)

┌─ process timing ─────────────────────────────────────┬─ overall results ─────┐
│        run time : 0 days, 0 hrs, 42 min, 52 sec      │  cycles done : 1502   │
│   last new path : none yet (odd, check syntax!)      │  total paths : 4      │
│ last uniq crash : none seen yet                      │ uniq crashes : 0      │
│  last uniq hang : none seen yet                      │   uniq hangs : 0      │
├─ cycle progress ────────────────────┬─ map coverage ─┴───────────────────────┤
│  now processing : 3* (75.00%)       │    map density : 6.82% / 6.82%         │
│ paths timed out : 0 (0.00%)         │ count coverage : 1.00 bits/tuple       │
├─ stage progress ────────────────────┼─ findings in depth ────────────────────┤
│  now trying : havoc                 │ favored paths : 1 (25.00%)             │
│ stage execs : 66/256 (25.78%)       │  new edges on : 1 (25.00%)             │
│ total execs : 1.54M                 │ total crashes : 0 (0 unique)           │
│  exec speed : 597.1/sec             │  total tmouts : 1 (1 unique)           │
├─ fuzzing strategy yields ───────────┴───────────────┬─ path geometry ────────┤
│   bit flips : 0/128, 0/124, 0/116                   │    levels : 1          │
│  byte flips : 0/16, 0/12, 0/4                       │   pending : 0          │
│ arithmetics : 0/896, 0/0, 0/0                       │  pend fav : 0          │
│  known ints : 0/88, 0/336, 0/176                    │ own finds : 0          │
│  dictionary : 0/0, 0/0, 0/0                         │  imported : n/a        │
│       havoc : 0/1.54M, 0/0                          │ stability : 100.00%    │
│        trim : 99.89%/67, 0.00%                      ├────────────────────────┘
└─────────────────────────────────────────────────────┘             [cpu:200%]

`

**Comment by str4d** [comment]
Concept ACK, and the changes look sensible. I'll review more closely when I'm at my laptop and can actually try compiling it.

File: .gitignore [hunk 1]
```diff
@@ -109,3 +109,7 @@ libzcashconsensus.pc
 
 contrib/debian/files
 contrib/debian/substvars
+
+src/fuzzing/*/output
+src/fuzz.cpp
+src/fuzz.h
```
Comment (str4d) [inline .gitignore]: It would be good to add these two copied files to `make clean`, so they aren't around during normal development.

**Comment by defuse** [comment]
Added a commit to have `make clean` delete fuzz.cpp.

**Comment by defuse** [comment]
There's a bug in this, AFL is trying to pass the data to stdin but the fuzzing stubs are trying to read from a file, that's why there's no new paths.

**Comment by defuse** [comment]
Fixed the bug in the latest commit.

**Comment by zebambam** [comment]
utACK

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
(What looks like a) transient gtest failure in one of the builders, so I manually cancelled. I'll retry shortly.

**Comment by str4d** [comment]
@zkbot retry

**Pull request merged**